homeassistant/components/sunsynkweb/sensor.py

```diff
@@ -0,0 +1,164 @@
+"""Sensors for the sunsynk web api."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import decimal
+import logging
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.const import PERCENTAGE, UnitOfEnergy, UnitOfPower
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import SunsynkConfigEntry
+from .coordinator import SunsynkUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class SunsynkEntitySensorDescription(SensorEntityDescription):
+    """A small wrapper for describing sunsynk entity sensors."""
+
+    agg_func: Callable = sum
+
+
+class SunSynkApiSensor(CoordinatorEntity[SunsynkUpdateCoordinator], SensorEntity):
+    """Parent class for sunsynk api exposing sensors.
+
+    The sensors expose the sum of the power across all inverters
+    for plants that have more than one inverter.
+    State of charge is normally shared between inverters, so this will take
+    the maximum of the state of charge.
+    """
+
+    _attr_has_entity_name = True
+
+    def __init__(self, coordinator, description) -> None:
+        """Initialise the common elements for sunsynk web api sensors."""
+        CoordinatorEntity.__init__(self, coordinator, context=coordinator)
+        self.coordinator = coordinator
+
+        self._attr_unique_id = (
+            f"{description.key}_{sum(p.id for p in coordinator.cache.plants)}"
+        )
+
+        self._attr_translation_key = description.translation_key
+        self.entity_description = description
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if self.coordinator.cache is not None:
+            self.native_value = self.entity_description.agg_func(
+                decimal.Decimal(getattr(i, self.entity_description.key))
+                for i in self.coordinator.cache.plants
+            )
+        else:
+            _LOGGER.debug("Not updating sensor as coordinator cache is empty. ")
+        self.async_write_ha_state()
+
+
+SENSOR_DESCRIPTIONS = [
```

Comment:
preference (and consistency with the rest of the project) please put everything from here on before the `SunSyncApiSensor` class